patches/minecraft/net/minecraft/block/Block.java.patch

```diff
@@ -1029,39 +1029,117 @@
 +    {
 +        IBlockState state = world.func_180495_p(pos);
 +        for (IProperty<?> prop : state.func_177228_b().keySet())
 +        {
 +            if (prop.func_177701_a().equals("facing") || prop.func_177701_a().equals("rotation"))
 +            {
-+                world.func_175656_a(pos, state.func_177231_a(prop));
-+                return true;
++                if(state.func_177231_a(prop).func_177229_b(prop) != state.func_177229_b(prop))
++                {
++                    return true;
++                }
 +            }
 +        }
 +        return false;
 +    }
-+
++    
++    /**
++     * Rotate the block around the specified axis, clockwise as looking from that side. Doesn't take into account
++     * the player and where on the block they're looking.<br/>
++     * Note: for mod blocks, this is up to the block and modder to decide, but it <b>IS</b> mandated that it be a rotation around the
++     * face, for compatibility reasons. If you know about the player and where they're looking at, implement
++     * {@link Block#rotateBlock(World, BlockPos, EnumFacing, EntityPlayer, RayTraceResult)}.
++     *
++     * @param world The world
++     * @param pos Block position in world
++     * @param axis The axis to rotate around
++     */
++    public void rotateBlockAround(World world, BlockPos pos, EnumFacing axis)
++    {
++        IBlockState state = world.func_180495_p(pos);
++        for (IProperty<?> prop : state.func_177228_b().keySet())
++        {
++            if (prop.func_177701_a().equals("facing") || prop.func_177701_a().equals("rotation"))
++            {
++                world.func_175656_a(pos, state.func_177231_a(prop));
++                return;
++            }
++        }
++    }
++    
++    /**
++     * Checks if the block can be rotated by the player. Takes into account
++     * the player and where on the block they're looking.<br/>
++     * Note: for mod blocks, its behavior is up to the block and modder to decide, but it is <b>NOT</b> mandated that it be a
++     * rotation around the face. If you don't know about the player or where they're looking at, implement
++     * {@link Block#canRotateBlockAround(World, BlockPos, EnumFacing)}.
++     *
++     * @param world The world
++     * @param pos Block position in world
++     * @param axis The axis to rotate around
++     * @param player The player that's attempting to rotate the block
++     * @param hit The position on the block where the player clicked
++     * @return True if it can be rotated, false if it can't
++     */
++    public boolean canPlayerRotate(World world, BlockPos pos, EnumFacing axis, EntityPlayer player, RayTraceResult hit)
++    {
++        return canRotateBlockAround(world, pos, axis);
++    }
++    
 +    /**
-+     * Get the rotations that can apply to the block at the specified coordinates. Null means no rotations are possible.
-+     * Note, this is up to the block to decide. It may not be accurate or representative.
```

Comment:
While talking about it on the FC channel we came to the conclusion that the current system is too abstract for most cases where the player is not the one rotating the block. That's why they've been split into different methods. Also, getValidRotations() had a purpose at first but when rotateBlock() was generalized it lost all its meaning and most mods didn't even check it before performing a rotation.
Now we have a method that ensures the block will be rotated around the specified axis for external rotation systems, as well as one that can be a bit more abstract for things that want to handle rotation differently with a wrench (or do not want a wrench to rotate the block, but they want frames to do it, for example - which has been requested for a while).